keps/sig-network/3836-kube-proxy-improved-ingress-connectivity-reliability/README.md

```diff
@@ -0,0 +1,346 @@
+the node.
+
+This KEP proposes two changes:
+
+1. That Kube-proxy's includes the `healthz` state AND its current answer w.r.t
+  the local endpoints, when it answers to the HC for class B) services.
+
+2. That Kube-proxy enables connection draining for terminating Nodes. This is to
+  be done by Kube-proxy by inspecting a field on the Node object which indicates
+  that the "node is terminating/deleting" and when seen, starts failing the HC.
+
+```
+<<[UNRESOLVED HC-linked-service-fields ]>>
+We need to agree on this field. `deletionTimestamp` is the perfect indicator for
+this, but....the Node object does not have a finalizer, so that wouldn't help.
+@thockin advocates for `.spec.unschedulable`, but I am inclined to say "that
+doesn't indicate if a Node is going away". `unschedulable` just says that no
+workloads can be assigned to the Node anymore. It is true that it's usually
+followed up by a drain...but I am not sure it's good to implement something
+which can easily kill ingress connectivity to a Node / cluster by simply doing
+`kubectl cordon node ...` (It would kill ingress connectivity to the entire
+cluster if the ingress controller is running on the node which is being cordoned)
+
+Maybe we need a KEP for a finalizer on the Node object (*troll face)
+<<[/UNRESOLVED]>>
+```
+
+```
+<<[UNRESOLVED HC-linked-service-fields ]>>
+Figure out which other cloud-providers configure the HC to target the service
+proxy for the HC
+<<[/UNRESOLVED]>>
+```
+
+## Motivation
+
+The motivation for each change is:
+
+1. Not including the `healthz` state can cause Kube-proxy to indicate to a load
+  balancer that it should send traffic to a Node simply because the endpoint is
+  scheduled there, even though Kube-proxy might not be healthy and successfully
+  managed to write the rules required for actually being able to forward traffic
+  to the endpoint.
+
+2. Workloads on terminating Nodes would benefit from connection draining as to
+  have their connections gracefully terminated. Workloads running on terminating
+  nodes get drained simultaneously with Kube-proxy today, this means that
+  there's no guarantee that connection draining will be enabled for them
+  currently. That might be the case if Kube-proxy is terminated before the
+  endpoint, but there's no guarantee of that. If Kube-proxy reacts to a "node is
+  about to go away" signal, then Kube-proxy can start prematurely failing the HC
+  and enable connection draining for these workloads.
+
+### Goals
+
+- Have kube-proxy accurately reflect its `healthz` state in its response to LB HCs
+  for class B) services.
+
+- Connection drain terminating Nodes
+
+### Non-Goals
+
+- Aligning cloud provider HCs for class A) services. Cloud providers like
+  Azure/AWS do not configure their HCs to point to the service proxy. Instead
+  they connect to the `NodePort` defined for the service. As to have them
+  benefit from the proposals of this KEP, they would need to change their
+  implementation.
```

Comment:
As I mentioned in a comment in the other KEP, it seems like right now there's no right/recommended way for cloud LBs to do health checks, and fixing that seems in scope. (Formally document somewhere, "If the cloud LB does X, then it may run into problem Y, so we recommend cloud providers configure it to do Z instead")

Reply:
I agree that it would be good to have an "implementers guide" for a cloud-provider.  I am not sure where that should live, though

Reply:
Should I mention this as a task for the KEP to go beta? Something like: "writing docs"?

Reply:
It could even be a page under https://kubernetes.io/docs/concepts/ titled: "Recommendations for cloud providers to better perform health checking on Kubernetes clusters". We can figure out the title later.

Reply:
SGTM - just a mention here is sufficient